formats/html.ts

```diff
@@ -21,24 +73,340 @@ export class HtmlImporter extends FormatImporter {
+			.map(file => this.processFile(folder, file)));
+		this.showResult({
+			total: results.map(({ total }) => total).reduce((left, right) => left + right, 0),
+			failed: results.map(({ failed }) => failed).flat(),
+			skipped: results.map(({ skipped }) => skipped).flat()
+		});
+	}
+
+	async processFile(folder: TFolder, file: PickedFile) {
+		const results: ImportResult = {
+			total: 1,
+			failed: [],
 			skipped: [],
-			failed: []
 		};
+		if (!(file instanceof NodePickedFile)) {
+			results.skipped.push(file.name);
+			return results;
+		}
+		let mdFile: TFile | null = null;
+		try {
+			let mdContent = htmlToMarkdown(await file.readText());
+			const pathURL = nodeUrl.pathToFileURL(file.filepath);
+			mdFile = await this.saveAsMarkdownFile(folder, file.basename, "");
+
+			const ast = [];
+			let read = 0;
+			let next = 0;
+			do {
+				next = mdContent.indexOf("!", read);
+				if (next === -1) {
+					next = mdContent.length;
+				}
+				const link = parseMarkdownLink(mdContent.slice(next), false);
+				if (link) {
+					ast.push(mdContent.slice(read, next), {
+						link,
+						text: mdContent.slice(next, next + link.read),
+					});
+					next += link.read;
+				} else {
+					next += "!".length;
+					ast.push(mdContent.slice(read, next));
+				}
+				read = next;
+			} while (read < mdContent.length)
+			const transformedAST = await Promise.all(ast
+				.map(async ast => {
+					if (typeof ast === "string") {
+						return { text: ast };
+					}
+					try {
+						let { link, link: { path: linkpath, display }, text } = ast;
+						const correctedPath = linkpath.startsWith("//") ? `https:${linkpath}` : linkpath;
+						let url;
+						try {
+							url = new URL(correctedPath);
+						} catch (e) {
+							if (!(e instanceof TypeError)) {
+								throw e;
+							}
+							url = new URL(normalizePath(correctedPath)
+								.split("/")
+								.map(encodeURIComponent)
+								.join("/"), pathURL);
+						}
+						const attachment = await this.downloadAttachmentCached(mdFile, url);
+						if (attachment instanceof TFile) {
+							text = this.app.fileManager.generateMarkdownLink(attachment, mdFile.path, "", display);
+						}
+						return { text, attachment, link } as const;
+					} catch (e) {
+						console.error(e);
+						return { text: ast.text };
+					}
+				}));
+			results.total += transformedAST
+				.filter(({ link }) => link)
+				.length;
+			results.failed = results.failed.concat(transformedAST
+				.filter(({ attachment }) => attachment === "failed")
+				.map(({ link: { path } }) => path));
+			results.skipped = results.skipped.concat(transformedAST
+				.filter(({ attachment }) => attachment === "skipped")
+				.map(({ link: { path } }) => path));
+
+			mdContent = transformedAST.map(({ text }) => text).join("");
+			await this.app.vault.modify(mdFile, mdContent);
+		} catch (e) {
+			console.error(e);
+			results.failed.push(file.toString());
+			if (mdFile) {
+				try {
+					await this.app.vault.delete(mdFile);
```

Comment:
> Hmm maybe it doesn't make sense to delete the file if any attachment download fails?

This does not happen because there is a `try-catch` block for downloading attachments: https://github.com/obsidianmd/obsidian-importer/pull/33/files/b1c500b454a1dfe1b5f40112f2faff455ac116b1#diff-3817194cf9ed55204dc4518448fc9ad1f788b08fec4b4d3aa16b8a989a28ba86R147-R150

For the updated code, `Promise.allSettled` is used instead, which achieves the above same result: https://github.com/obsidianmd/obsidian-importer/pull/33/files#diff-3817194cf9ed55204dc4518448fc9ad1f788b08fec4b4d3aa16b8a989a28ba86R97

> Perhaps we should not save to file immediately after converting to md, and instead only save when everything has been downloaded? I know we'll need a path for link generation, but since it's only used for relative link resolution, I think we can generate a fake one of where we think the file will end up at.

The undocumented method `getAvailablePathForAttachments` needs a `TFile` as the third argument: https://github.com/obsidianmd/obsidian-importer/pull/33/files#diff-3817194cf9ed55204dc4518448fc9ad1f788b08fec4b4d3aa16b8a989a28ba86R241-R242

Reply:
Ah your are correct for both!